lib/readline.js

```diff
@@ -474,8 +491,60 @@ Interface.prototype._insertString = function(c) {
     // a hack to get the line refreshed if it's needed
     this._moveCursor(0);
   }
+
+  searchHistory.call(this);
 };
 
+function appendSearchResult(result) {
+  // this.previewResult = result;
+
+  // Cursor to left edge.
+  cursorTo(this.output, 0);
+  clearScreenDown(this.output);
+
+  // Based on the line and match result
+  // write the data
+  if (result !== '') {
+    this.output.write(buildReverseSearchPrompt(this.line, result));
+    cursorTo(this.output, this.cursor + kReverseSearchPrompt.length - 2);
+  } else if (this.line === '') {
+    this.output.write(buildReverseSearchPrompt());
+    cursorTo(this.output, this.cursor + kReverseSearchPrompt.length - 2);
+  } else {
+    this.output.write(buildReverseSearchPrompt(this.line, ''));
+    cursorTo(this.output, this.cursor + kFailedReverseSearchPrompt.length);
+  }
+
+}
+
+function searchText() {
+  let result = '';
+  const historySet = new Set(this.history);
+  for (;this.reverseSearchIndex < [...historySet].length;
```

Comment:
There is no need to convert the set back to an array. Just use `.size` instead. The same below.

However, this approach does not work in all cases. The unique set could have less entries than the original history and the entry to look for might be at the end of the history which would not be reached here anymore.

E.g.: The set of `[1, 1, 1, 2]` is `{ 1, 2 }` and if we look for `2` it would only check history entry 1 or 2 and would give up afterwards.

I personally would just iterate directly over the array.

```js
function searchHistory(self, unique, length) {
  while (++self.reverseSearchIndex < length) {
    const entry = self.history[self.reverseSearchIndex];
    if (unique.has(entry) continue;
    unique.add(entry);
    const start = entry.indexOf(self.line)
    if (start !== -1) {
	  const output = `(reverse-i-search)\`${self.line}': `;
	  self.output.write(`${output}${entry}`);
	  cursorTo(self.output, output.length + start);
      return true;
    }
  }
}

function reverseSearch(self) {
  const unique = new Set();
  const start = self.reverseSearchIndex;
  let matched = searchHistory(self, unique, this.history.length);
  if (matched)
    return;
  self.reverseSearchIndex = 0;
  matched = searchHistory(self, unique, start);
  if (matched)
    return;
  self.reverseSearchIndex = -1;
  const output = `${kFailedReverseSearchPrompt}${text}':`;
  self.output.write(output);
  cursorTo(self.output, output.length);
}

// Further below
       case 'r':
        if (this.reverseSearchIndex !== -2)
          this.reverseSearchIndex = -1;
        reverseSearch(this);

// Further below
      case 'return':  // carriage return, i.e. \r
        this._sawReturnAt = Date.now();
        if (this.reverseSearchIndex !== -2) {
          if (this.reverseSearchIndex !== -1)
            this.line = this.history[this.reverseSearchIndex];
          this.reverseSearchIndex = -2;
        }
```

The code is untested but it should roughly work like that.

Btw: the bash would change the `historyIndex` which this implementation does not but I think it's fine as it is.